clients/da-vinci-client/src/main/java/com/linkedin/davinci/storage/StorageService.java

```diff
@@ -366,18 +366,33 @@ public synchronized StorageEngine openStore(
 
     LOGGER.info("Creating/Opening Storage Engine {} with type: {}", topicName, storeConfig.getStorePersistenceType());
     StorageEngineFactory factory = getInternalStorageEngineFactory(storeConfig);
-    engine =
+    StorageEngine newEngine =
         factory.getStorageEngine(storeConfig, isReplicationMetadataEnabled(topicName, factory.getPersistenceType()));
-    engine.updateStoreVersionStateCache(initialStoreVersionStateSupplier.get());
-    storageEngineRepository.addLocalStorageEngine(engine);
+    newEngine.updateStoreVersionStateCache(initialStoreVersionStateSupplier.get());
+
+    // Let's check if a previous incarnation of the storage engine existed earlier
+    ReferenceCounted<DelegatingStorageEngine> refCountedStorageEngine =
+        this.storageEngines.compute(topicName, (k, v) -> {
+          if (v != null) {
+            v.get().setDelegate(newEngine);
+            LOGGER.info(
+                "Injected newly created storage engine into existing ref-counted delegating storage engine for {}.",
+                k);
+          }
+          return v;
+        });
+    DelegatingStorageEngine delegatingStorageEngine =
+        refCountedStorageEngine == null ? new DelegatingStorageEngine(newEngine) : refCountedStorageEngine.get();
```

Comment:
I have a question about this - should we add a ReferenceCounted<DelegatingStorageEngine> refCountedStorageEngine to the above Map<String, ReferenceCounted<DelegatingStorageEngine>> storageEngines? Then the bottom method getRefCountedStorageEngine is no longer needed? Maybe it is the same but just have this question why don't we update the map "cache" right away